iterator-test-harness/src/main/java/org/apache/accumulo/iteratortest/testcases/DeepCopyTestCase.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.accumulo.iteratortest.testcases;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.TreeMap;
+
+import org.apache.accumulo.core.data.ByteSequence;
+import org.apache.accumulo.core.data.Key;
+import org.apache.accumulo.core.data.Value;
+import org.apache.accumulo.core.iterators.SortedKeyValueIterator;
+import org.apache.accumulo.iteratortest.IteratorTestInput;
+import org.apache.accumulo.iteratortest.IteratorTestOutput;
+import org.apache.accumulo.iteratortest.IteratorTestUtil;
+import org.apache.accumulo.iteratortest.environments.SimpleIteratorEnvironment;
+
+/**
+ * Test case that verifies that an iterator can use the generated instance from {@code deepCopy}.
+ */
+public class DeepCopyTestCase extends OutputVerifyingTestCase {
+
+  @Override
+  public IteratorTestOutput test(IteratorTestInput testInput) {
+    final SortedKeyValueIterator<Key,Value> skvi = IteratorTestUtil.instantiateIterator(testInput);
+    final SortedKeyValueIterator<Key,Value> source = IteratorTestUtil.createSource(testInput);
+
+    try {
+      skvi.init(source, testInput.getIteratorOptions(), new SimpleIteratorEnvironment());
+
+      SortedKeyValueIterator<Key,Value> copy = skvi.deepCopy(new SimpleIteratorEnvironment());
```

Comment:
> is the purpose to test only system iterators, as opposed to generic tests to check user-defined iterators? That makes more sense. 

Yep. First and foremost, my goal was to provide a framework for users to test iterators without running them through Accumulo (since it's often hard to know when you're hitting some of the edge cases). However, I can totally see it growing into a framework to catch "common pitfalls" like you outlined :)

> Some users may define iterators that expect a column family, whereas all the tests seek with Collections.<ByteSequence> emptySet(), false

Yeah, this is something I'm not quite sure how to handle colfams (I'm throwing that into the same bucket as the compaction stuff and more that Keith pointed out).

>  If we make this public API, we should state that user iterators do not necessarily need to pass all the IteratorTestCases, or at least that failing an IteratorTestCase does not imply a bug.

Very good point. We don't really have an "evolving" annotation. Maybe we just talk about it and the lack of public API declares it as evolving intrinsically?